frame/transaction-payment/src/payment.rs

```diff
@@ -0,0 +1,124 @@
+///! Traits and default implementation for paying transaction fees.
+use crate::{NegativeImbalanceOf, Trait};
+use codec::FullCodec;
+use frame_support::{
+	traits::{Currency, ExistenceRequirement, Get, Imbalance, OnUnbalanced, WithdrawReason},
+	unsigned::TransactionValidityError,
+};
+use sp_runtime::{
+	traits::{AtLeast32BitUnsigned, DispatchInfoOf, MaybeSerializeDeserialize, PostDispatchInfoOf, Saturating, Zero},
+	transaction_validity::InvalidTransaction,
+};
+use sp_std::{fmt::Debug, marker::PhantomData};
+
+/// Handle withdrawing, refunding and depositing of transaction fees.
+pub trait OnChargeTransaction<T: Trait> {
+	/// The currency type in which fees will be paid.
+	type Balance: AtLeast32BitUnsigned + FullCodec + Copy + MaybeSerializeDeserialize + Debug + Default;
+	type LiquidityInfo: Default;
+
+	/// Before the transaction is executed the payment of the transaction fees
+	/// need to be secured.
+	fn withdraw_fee(
+		who: &T::AccountId,
+		call: &T::Call,
+		dispatch_info: &DispatchInfoOf<T::Call>,
+		fee: Self::Balance,
+		tip: Self::Balance,
+	) -> Result<Self::LiquidityInfo, TransactionValidityError>;
+
+	/// After the transaction was executed and the correct fees where collected,
+	/// the resulting funds can be spend here.
+	fn deposit_fee(
+		who: &T::AccountId,
+		dispatch_info: &DispatchInfoOf<T::Call>,
+		post_info: &PostDispatchInfoOf<T::Call>,
+		fee: Self::Balance,
+		tip: Self::Balance,
+		liquidity_info: Self::LiquidityInfo,
+	) -> Result<(), TransactionValidityError>;
```

Comment:
```suggestion
	/// After the transaction was executed the correct fees can be calculated.
	/// This function should refund any overpaid fees and optionally deposit
	/// the adjusted amount.
	///
	/// Note: The `fee` already includes the `tip`.
	fn adjust_and_deposit_fee(
		who: &T::AccountId,
		dispatch_info: &DispatchInfoOf<T::Call>,
		post_info: &PostDispatchInfoOf<T::Call>,
		adjusted_fee: Self::Balance,
		tip: Self::Balance,
		already_withdrawn: Self::LiquidityInfo,
	) -> Result<(), TransactionValidityError>;
```

Reply:
I applied your changes in 623b449e2ab1fcb7bd5c7fa1d871e99fdf0cff01. Thanks for reviewing ! 👍 💯